### PR TITLE
Fix for sparsity-preserving clustering

### PR DESCRIPTION
**Pull request by TamasArm**
* small fix to stop centroids from drifting
* unit test coverage for fix

Summary of change:
Ensure the sparsity mask is updated during training time in sparsity-preserving clustering. Due to cluster-weight associations updates, the location of zero weights may change during training. This patch updated the sparsity mask according to such changes. It also adds extra tests to ensure sparsity aware clustering works during distributed training.

### CODE REVIEW

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py [hunk 1]
```diff
@@ -215,6 +215,10 @@ def update_clustered_weights_associations(self):
               pulling_indices, original_weight))
 
       if self.preserve_sparsity:
+        # Re-discover the sparsity masks to avoid drifting
+        self.sparsity_masks[weight_name] = (
+            tf.cast(tf.math.not_equal(clustered_weights, 0), dtype=tf.float32)
+        )
```
Comment (wwwind) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py:221]: It looks to me that this could be wrong: we form mask here from clustered_weights and then we multiply them with this mask.

In clustering we have trainable: weights and centroids. They could move away from zeros during training. To keep sparsity, we need to 1. to apply original sparsity_mask that is built in the build function to the original_weights; 2. to assign the smallest centroid back to zero, if there are no zero centroids. Here there is a variation: we could let original_weights to train as they want, because we use them to compute gradients in the backpropagation and give them more freedom, but we need to apply sparsity mask before we compute pulling indices. However, I would update with the original sparsity_mask the original_weights during training.

**Review by wwwind — APPROVED** [review]

**Review by akarmi — APPROVED** [review]

**Comment by TamasArm** [comment]
@daverim Hi David, could you please take a look at this and retrigger the pull request labeler if possible? Thank you in advance!

**Review by daverim — APPROVED** [review]
